pex/resolver.py

```diff
@@ -610,6 +611,90 @@ def build_wheels(
+        # type: (...) -> DirectRequirements
+
+        # 3. All requirements are now in wheel form: calculate any missing direct requirement
+        #    project names from the wheel names.
+        with TRACER.timed(
+            "Calculating project names for direct requirements:"
+            "\n  {}".format("\n  ".join(map(str, direct_requirements)))
+        ):
+
+            def iter_direct_requirements():
+                # type: () -> Iterator[Requirement]
+                for requirement in direct_requirements:
+                    if not isinstance(requirement, LocalProjectRequirement):
+                        yield requirement.requirement
+                        continue
+
+                    install_reqs = install_requests.get(requirement.path)
+                    if not install_reqs and local_project_directory_to_sdist:
+                        local_project_directory = local_project_directory_to_sdist.get(
+                            requirement.path
+                        )
+                        if local_project_directory:
+                            install_reqs = install_requests.get(local_project_directory)
+                    if not install_reqs:
+                        raise AssertionError(
+                            "Failed to compute a project name for {requirement}. No corresponding "
+                            "wheel was found from amongst:\n{install_requests}".format(
+                                requirement=requirement,
+                                install_requests="\n".join(
+                                    sorted(
+                                        "{path} -> {wheel_path} {fingerprint}".format(
+                                            path=path,
+                                            wheel_path=build_result.wheel_path,
+                                            fingerprint=build_result.fingerprint,
+                                        )
+                                        for path, build_results in install_requests.items()
+                                        for build_result in build_results
+                                    )
+                                ),
+                            )
+                        )
+                    for install_req in install_reqs:
+                        yield requirement.as_requirement(dist=install_req.wheel_path)
+
+            direct_requirements_by_project_name = defaultdict(
+                OrderedSet
+            )  # type: DefaultDict[ProjectName, OrderedSet[Requirement]]
+            for direct_requirement in iter_direct_requirements():
+                direct_requirements_by_project_name[direct_requirement.project_name].add(
+                    direct_requirement
+                )
+            return cls(direct_requirements_by_project_name)
+
+    _direct_requirements_by_project_name = (
+        attr.ib()
+    )  # Mapping[ProjectName, OrderedSet[Requirement]]
+
+    def adjust(self, distributions):
```

Comment:
What's this adjusting? (I acknowledge it's just moved code.)

Reply:
The name of the extracted class is indicative here, it adjusts a distribution to note the direct top level requirement it satisfies, if any. Most dists usually don't get adjusted since they satisfy only transitive requirements. The need for associating dists to the direct requirements they satisfy goes back to handling local project directory dependencies. These have no project name known except after the resolve process which builds them into a wheel with a name and version.